amplitude.py

```diff
@@ -1,20 +1,79 @@
+import boto.s3
+import hashlib
+import hmac
 import json
+import os
+import requests
 import sys
 
-def transform(line):
-    try:
-        parsed_line = json.loads(line)
+from threading import Timer
 
-        assert(type(parsed_line) is dict)
+AMPLITUDE_API_KEY = os.environ["FXA_AMPLITUDE_API_KEY"]
+AWS_ACCESS_KEY = os.environ["FXA_AWS_ACCESS_KEY"]
+AWS_SECRET_KEY = os.environ["FXA_AWS_SECRET_KEY"]
+HMAC_KEY = os.environ["FXA_AMPLITUDE_HMAC_KEY"]
 
-        event = parsed_line["Fields"]
+# For crude pre-emptive rate-limit obedience.
+MAX_EVENTS_PER_BATCH = 10
+MAX_BATCHES_PER_SECOND = 100
+BATCH_INTERVAL = 1.0 / MAX_BATCHES_PER_SECOND
 
+def handle (message):
+    # http://docs.aws.amazon.com/AmazonS3/latest/dev/notification-content-structure.html
+    records = json.loads(message)["Records"]
+    for record in records:
+        if record["eventSource"] != "aws:s3":
+            continue
```

Comment:
Currently, my crude rate-limiting workaround is dependent on there only being one record per message. But if we need to, I can expand it to sanely handle multiple records.

Possibly I should `assert(len(records) == 1)` to make this assumption explicit, but I wasn't sure if there might legitimately be records for other event sources in the message too?

Reply:
I had second thoughts on this and decided to assume/assert 1 record after all. Let me know if that turns out to be wrong, I'm happy to change it back to a loop.

Reply:
The switch to `time.sleep` in 91501be meant I could reinstate this.